---
author-meta:
- James A. Fellows Yates
- Thiseas C. Lamnidis
- Maxime Borry
- "Aida Andrades Valtue\xF1a"
- "Zandra Fagern\xE4s"
- Stephen Clayton
- Maxime U. Garcia
- Judith Neukamm
- Alexander Peltzer
bibliography:
- content/manual-references.json
date-meta: '2020-05-29'
header-includes: "<!--\nManubot generated metadata rendered from header-includes-template.html.\nSuggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html\n-->\n<meta name=\"dc.format\" content=\"text/html\" />\n<meta name=\"dc.title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"citation_title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"og:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"twitter:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"dc.date\" content=\"2020-05-29\" />\n<meta name=\"citation_publication_date\" content=\"2020-05-29\" />\n<meta name=\"dc.language\" content=\"en-GB\" />\n<meta name=\"citation_language\" content=\"en-GB\" />\n<meta name=\"dc.relation.ispartof\" content=\"Manubot\" />\n<meta name=\"dc.publisher\" content=\"Manubot\" />\n<meta name=\"citation_journal_title\" content=\"Manubot\" />\n<meta name=\"citation_technical_report_institution\" content=\"Manubot\" />\n<meta name=\"citation_author\" content=\"James A. Fellows Yates\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5585-6277\" />\n<meta name=\"twitter:creator\" content=\"@jfy133\" />\n<meta name=\"citation_author\" content=\"Thiseas C. Lamnidis\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-4485-8570\" />\n<meta name=\"twitter:creator\" content=\"@TCLamnidis\" />\n<meta name=\"citation_author\" content=\"Maxime Borry\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-9140-7559\" />\n<meta name=\"twitter:creator\" content=\"@notmaxib\" />\n<meta name=\"citation_author\" content=\"Aida Andrades Valtue\xF1a\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"XXXX-XXXX-XXXX-XXXX\" />\n<meta name=\"twitter:creator\" content=\"@aidaanva\" />\n<meta name=\"citation_author\" content=\"Zandra Fagern\xE4s\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2667-3556\" />\n<meta name=\"twitter:creator\" content=\"@ZandraSelina\" />\n<meta name=\"citation_author\" content=\"Stephen Clayton\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History\" />\n<meta name=\"citation_author_orcid\" content=\"XXXX-XXXX-XXXX-XXXX\" />\n<meta name=\"citation_author\" content=\"Maxime U. Garcia\" />\n<meta name=\"citation_author_institution\" content=\"Department of Oncology-Pathology, Karolinska Institutet\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2827-9261\" />\n<meta name=\"twitter:creator\" content=\"@gau\" />\n<meta name=\"citation_author\" content=\"Judith Neukamm\" />\n<meta name=\"citation_author_institution\" content=\"Palaeogenetics Group, Institute of Evolutionary Medicina, University of Zurich\" />\n<meta name=\"citation_author_orcid\" content=\"XXXX-XXXX-XXXX-XXXX\" />\n<meta name=\"twitter:creator\" content=\"@JudithNeukamm\" />\n<meta name=\"citation_author\" content=\"Alexander Peltzer\" />\n<meta name=\"citation_author_institution\" content=\"Translational Medicine &amp; Clinical Pharmacology, Boehringer Ingelheim Pharma GmbH &amp; CO KG, Biberach an der Riss\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-6503-2180\" />\n<meta name=\"twitter:creator\" content=\"@alex_peltzer\" />\n<link rel=\"canonical\" href=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"og:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"twitter:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_fulltext_html_url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_pdf_url\" content=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"application/pdf\" href=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"text/html\" href=\"https://apeltzer.github.io/eager2-paper/v/0b87d41c8a2654c7d71707b22866ea596bc9c919/\" />\n<meta name=\"manubot_html_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/0b87d41c8a2654c7d71707b22866ea596bc9c919/\" />\n<meta name=\"manubot_pdf_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/0b87d41c8a2654c7d71707b22866ea596bc9c919/manuscript.pdf\" />\n<meta property=\"og:type\" content=\"article\" />\n<meta property=\"twitter:card\" content=\"summary_large_image\" />\n<link rel=\"icon\" type=\"image/png\" sizes=\"192x192\" href=\"https://manubot.org/favicon-192x192.png\" />\n<link rel=\"mask-icon\" href=\"https://manubot.org/safari-pinned-tab.svg\" color=\"#ad1457\" />\n<meta name=\"theme-color\" content=\"#ad1457\" />\n<!-- end Manubot generated metadata -->"
keywords:
- nf-core
- nextflow
- ancientDNA
- aDNA
- palaeogenetics
- archaeogenetics
- bioinformatics
lang: en-GB
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager
...






<small><em>
This manuscript
([permalink](https://apeltzer.github.io/eager2-paper/v/0b87d41c8a2654c7d71707b22866ea596bc9c919/))
was automatically generated
from [apeltzer/eager2-paper@0b87d41](https://github.com/apeltzer/eager2-paper/tree/0b87d41c8a2654c7d71707b22866ea596bc9c919)
on May 29, 2020.
</em></small>

## Authors



+ **James A. Fellows Yates**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5585-6277](https://orcid.org/0000-0001-5585-6277)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jfy133](https://github.com/jfy133)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [jfy133](https://twitter.com/jfy133)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Thiseas C. Lamnidis**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4485-8570](https://orcid.org/0000-0003-4485-8570)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [TCLamnidis](https://github.com/TCLamnidis)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [TCLamnidis](https://twitter.com/TCLamnidis)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Maxime Borry**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-9140-7559](https://orcid.org/0000-0001-9140-7559)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [maxibor](https://github.com/maxibor)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [notmaxib](https://twitter.com/notmaxib)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Aida Andrades Valtueña**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [XXXX-XXXX-XXXX-XXXX](https://orcid.org/XXXX-XXXX-XXXX-XXXX)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [aidaanva](https://github.com/aidaanva)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [aidaanva](https://twitter.com/aidaanva)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Zandra Fagernäs**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2667-3556](https://orcid.org/0000-0003-2667-3556)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ZandraFagernas](https://github.com/ZandraFagernas)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [ZandraSelina](https://twitter.com/ZandraSelina)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Stephen Clayton**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [XXXX-XXXX-XXXX-XXXX](https://orcid.org/XXXX-XXXX-XXXX-XXXX)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [sc13-bioinf](https://github.com/sc13-bioinf)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History
     · Funded by Max Planck Society
  </small>

+ **Maxime U. Garcia**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2827-9261](https://orcid.org/0000-0003-2827-9261)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [MaxUlysse](https://github.com/MaxUlysse)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [gau](https://twitter.com/gau)<br>
  <small>
     Department of Oncology-Pathology, Karolinska Institutet
     · Funded by Barncancerfonden
  </small>

+ **Judith Neukamm**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [XXXX-XXXX-XXXX-XXXX](https://orcid.org/XXXX-XXXX-XXXX-XXXX)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [JudithNeukamm](https://github.com/JudithNeukamm)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [JudithNeukamm](https://twitter.com/JudithNeukamm)<br>
  <small>
     Palaeogenetics Group, Institute of Evolutionary Medicina, University of Zurich
  </small>

+ **Alexander Peltzer**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-6503-2180](https://orcid.org/0000-0002-6503-2180)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [apeltzer](https://github.com/apeltzer)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [alex_peltzer](https://twitter.com/alex_peltzer)<br>
  <small>
     Translational Medicine & Clinical Pharmacology, Boehringer Ingelheim Pharma GmbH & CO KG, Biberach an der Riss
  </small>



## Abstract {.page_break_before}

The broadening utilization of ancient DNA (aDNA) to address anthropological,
palaeontological and biological questions is resulting in a rising diversity in
the size of laboratories and scale of analyses being performed. In the context
of this heterogeneous landscape, we present
[nf-core/eager](https://github.com/nf-core/eager), an advanced and entirely
redesigned pipeline for the analysis of ancient DNA genomic data. nf-core/eager
builds on existing ideas and concepts introduced in the original EAGER
pipeline @doi:10.1186/s13059-016-0918-z, and improves various aspects of the
analysis procedure by building on computational frameworks such as
Nextflow @doi:10.1038/nbt.3820 and nf-core @doi:10.1038/s41587-020-0439-x.
The pipeline aims to address three main points: adaptability to different
computing configurations, reproducibility to ensure robust analytical standards
in the field, and bring the pipeline up-to-date with the latest routine and ancient
genomic practises. This new version of EAGER has been developed within the
nf-core initiative, to ensure high quality software development and maintenance
support; contributing to a long-term lifecycle for the pipeline. nf-core/eager
will assist in ensuring that ancient DNA sequencing data can be utilised by a diverse range of
research groups and fields.


## Introduction

Ancient DNA has become a widely accepted source of biological data, helping
to provide new perspective for a range of fields including archaeology, ecology,
cultural heritage, and palaeontology. The utilisation of
next-generation-sequencing has allowed the recovery of ancient DNA from a
wide variety of sources, including but not limited to, the skeletal remains of
animals (@doi:10.1016/j.cub.2015.04.007, @doi:10.1038/nature12323,
@doi:10.1073/pnas.1901169116, @10.1073/pnas.1710186114), modern and archaic
humans (@doi:10.1038/s41586-018-0094-2, @doi:0.1038/s41586-018-0455-x), bacteria
(@doi:10.1038/nature13591, @doi:10.1073/pnas.1812865115,
@doi:10.1371/journal.ppat.1006997), Viruses (@doi:10.1038/s41586-018-0097-z,
@doi:10.7554/eLife.36666), plants (@doi:10.1111/eva.12594,
@doi:10.1038/s41559-019-0921-3), but also coprolites
(@doi:10.1016/j.chom.2019.08.018, @doi:10.7717/peerj.9001), dental calculus
(@doi:10.1038/ng.2906, @doi:10.1038/nature21674), sediments
(@doi:10.1038/nature12921, @doi:10.1126/science.aam9695), medical slides
(@doi:10.1093/molbev/msz264) parchment (@doi:10.1098/rstb.2013.0379), and most
recently ancient ‘chewing gum’
(@doi:10.1038/s41467-019-13549-9, @doi:10.1038/s42003-019-0399-1). Improvement
in laboratory protocols to increase yields of otherwise trace amount of DNA has
at the same time led to studies that can range from one to thousands of ancient
individuals (@doi:10.1038/nature25738, @doi10.1038/nature25778), spanning
single (@doi:10.1038/nature10549) to thousands of organisms
(@doi:10.1038/ng.2906). These differences of disciplines have led to a
heterogeneous landscape in terms of types of analyses thus types of computing
resources required by different labs - particularly in regions of the world
where ancient DNA is still emerging.

Two previously published and commonly used pipelines in the field are
PALEOMIX (@doi:10.1038/nprot.2014.063) and EAGER
(@doi:10.1186/s13059-016-0918-z). These two pipelines take a similar approach
to link together standard tools used for Illumina NGS data processing (quality
control, adapter removal/and or merging, mapping, genotyping etc.), but with a
specific focus on tools that are designed for or well-suited for ancient DNA
(such as bwa aln (@doi:10.1093/bioinformatics/btp324) for ultra-short reads and
mapDamage (@doi:10.1093/bioinformatics/btr347) for aDNA characteristic
evaluation). Yet, neither have had major updates to bring in line with current
routinely carried out aDNA analyses. In particular, metagenomic screening of
off-target genomic reads for pathogens or microbiomes
(@doi:10.1038/ng.2906, @doi:10.1038/nature21674) has becoming particularly
common, given it's role in revealing widespread infectious disease and possible
epidemics that had previously been undetected in the archaeological
record (@doi:10.1016/j.cell.2015.10.009, @doi:10.1016/j.cub.2017.10.025,
@doi:10.1038/s41586-018-0097-z, @doi:10.7554/eLife.36666). Without easy access
to the latest field-established analytical routines, aDNA studies from groups
new to the field risk being published without the necessary quality controls
checks to ensure authenticity of their data and without yielding the full range
of possibilities from the data.

To address these short comings, we have completely re-implemented the latest
version of the EAGER pipeline in Nextflow, a domain-specific-language (DSL)
designed for the construction of ‘omic analysis pipelines
(@doi:10.1038/nbt.3820). In addition, the newly named pipeline - nf-core/eager -
has been developed in the context of the nf-core community framework
(@doi:10.1038/s41587-020-0439-x). nf-core enforces strict guidelines for
best-practises in software development to ensure robust, long-term maintenance
and high quality pipelines.


# Results and Discussion

## Scalability, Portability, and Efficiency

The reimplementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager
into various schedulers in POSIX High-Performance-Cluster (HPC) environments as
well as local computers. This portability allows both small and big labs to run
nf-core/eager regardless of the type of computer or cluster cluster sizes with
minimal effort or configuration, facilitating reproducibility and therefore
maintenance of standards within the field. This is further assisted by the
in-built compatibility with software environments and contains such as conda,
docker and singularity. This ensures exact versions of software are used by
a user, regardless of the set-up of the cluster environment. Another major
change with nf-core/eager is that the GUI input is now replaced with a
command-line-interface as the primary user interaction mode. This is more
compatible and portable with most HPCs (that may not offer X11 forwarding) and
is in line with the vast majority of bioinformatic tools. We therefore believe
this will not be a hindrance to new researchers from outside computational
biology, however the nf-core initiative is in the processing in the near future
of offering multiple alternatives including a command-line wizard, and a
web-based input GUI.

Reproducibility is further made easier through the use of 'profiles' that can
define configuration parameters for both the computing environment (schedulers,
cache locations, maximum resource etc.), which can be centrally managed to
ensure all users of a group use the same settings. Pipeline level profiles,
specifying default parameters for nf-core/eager itself is therefore more
portable between different groups through separation of the parameters from
input data. Compared to the original EAGER that utilised per-FASTQ XML files
with hardcoded filepaths, researchers can publish alongside their publications
the specific profile used in their nf-core/eager runs to ensure other groups can
generate the same results. Using of profiles also reduces mistakes caused by
insufficient 'prose' based reporting of program settings, which often occurs
in papers written by researchers unfamiliar with informatics. The default
nf-core/eager profile uses parameters evaluated in different aDNA specific
contexts (e.g. @doi:10.1186/1471-2164-13-178), and will be updated in each new 
release as new studies are published.

nf-core/eager provides improved efficiency over the original EAGER pipeline by
replacing the sample-by-sample sequential processing with Nextflow's
asynchronous parallelisation. This combined with nf-core/eager-defined
per-process customisation of resource parameters reduces unnecessary resource
allocation that can occur with new users to each step of an NGS data processing
pipeline. This is particularly pertinent given the increasing use of
centralised HPCs or cloud computing, which often use per-hour cost calculations.

## Latest aDNA practices

nf-core/eager follows a similar structural foundation with the original EAGER.
Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file.

This can be split into four main stages:

**ADD CITATIONS**

1. Preprocessing
   1. Sequencing quality control (FastQC)
   2. Sequencing artefact clean-up (AdapterRemoval2)
   3. Preprocessing statistics generation
2. Mapping and post- processing
   1. Alignment against reference genome (BWA, CircularMapper)
   2. Mapping quality filtering (Samtools)
   3. PCR duplicate removal (DeDup, Samtools MarkDuplicates)
   4. Mapping statistics generation (PreSeq, Qualimap)
3. aDNA Evaluation and Modification
   1. Damage profiling (DamageProfiler)
   2. aDNA reads selection (PMDTools)
   3. Damage removal (Bamutils)
   4. (Human) contamination estimation (ANGSD)
4. Genotyping and Consensus Sequencing (GATK, VCF2Genome)

In nf-core/eager, all tools also originally used in EAGER have been updated
to latest compatible versions, as available on conda to ensure widespread
accessibility and stability. **Modules that have not been carried over are
Schmutzi and MapDamage because ALEX TO WRITE SOMETHING NICE HERE**. Support
for Bowtie2 will be added in the near future, after consultation with the
nf-core user community. New tools to the basic workflow include fastp
(@doi:10.1093/bioinformatics/bty560) for the removal of poly-G sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms). We have also included the
FreeBayes genotyper as an alternative to the human-focused GATK tools. We have
also maintained the possibility of using the now unsupported GATK 
UnifiedGenotyper, as the GATK HaplotypeCaller performs _de novo_ assembly
around possible variants, which may not be suitable for low-coverage aDNA data.  

We have further extended functionality of the pipeline, primarily focusing on
allowing for ancient metagenomic analysis to be run alongside genomic of the
reference genome. We have added the ability to screen all off-target reads
from mapping to the supplied reference FASTA with two metagenomic profilers:
MALT (@doi:10.1101/050559, @doi:10.1038/s41559-017-0446-6) and Kraken2
(@doi:10.1186/s13059-019-1891-0). Characterisation of properties of authentic
ancient DNA (@doi:10.1073/pnas.0704665104) from MALT alignments is carried out with the HOPS pipeline
(@doi:10.1186/s13059-019-1903-0). The pipeline also integrates a SNP alignment
generation tool, MultiVCFAnalyzer (doi:10.1038/nature13591), which allows
assessment of level of cross-mapping to a reference genome from different
related taxa - a common challenge in ancient bacteria genome reconstruction
(@doi:10.1146/annurev-genom-091416-035526). Simple coverage statistics of
particular annotations (e.g. genes) of an input reference  is offered by bedtools (@doi:10.1093/bioinformatics/btq033).

New tools supplied with the pipeline include a dedicated 'endogenous DNA'
calculator (endorS.py, by A.A.V), to give approximate level of
on-target DNA yield within a sample. Secondly, ancient metagenomic studies may
include comparative samples from living day individuals
(@doi:10.1186/s40168-019-0717-3). To support open data, whilst respecting data
privacy, nf-core/eager includes a 'strip_fastq' script (by M.B.) which creates
raw FASTQ files but with reference-genome mapped reads removed. This then
allows safe upload of sequencing data to public repositories with identifiable
data removed. When using a human reference genome, nf-core/eager also can give
estimates of the biological sex of a given individual with Sex.DetEERRmine
(@doi:10.1038/s41467-018-07483-5).

Complex sequencing strategies is also now accounted for with nf-core/eager.
Given the large amount of sequencing often required to yield sufficient genome
coverage from ancient DNA data, palaeogeneticists can sometimes use multiple
(differently treated) libraries or sequencing runs. As an alternative to direct
paths, the pipeline can also accept a TSV file which includes file paths and
additional metadata such as sample name, library name, sequencing  lane, colour
chemistry and UDG treatment. This allows simultaneous processing and appropriate
merging of heterogeneous data of multiple sequencing runs and/or libraries
types.

Finally, the original EAGER tabular report format has been replaced with MultiQC
(@doi:10.1093/bioinformatics/btw354). The original EAGER pipeline required users
to look through many independent output directories and files to make full
assessment of their sequencing data. Aggregation of all log files into a single
interactive report will assist users in making fuller assessment of their
sequencing and analysis runs. Most tools within nf-core/eager have a
corresponding MultiQC module to ensure as complete evaluation as possible.

## Accessibility

Alongside the portable new pipeline report, we have written extensive documentation
on all parts of running and interpreting the output of the pipeline. Given that
a large fraction of aDNA researchers come from fields such as social sciences
that have limited computational training (such as J.A.F.Y), we have written
documentation that also gives guidance on how to interpret each section of the
report, specifically in the context of NGS and aDNA. This includes schematic
images (by Z.F.) of best practices or expected output, published under CC-BY
licenses to allow for use in other training material. This open-access resource
will make the aDNA discipline more accessible to researchers new to the field,
by providing practical and 'applied' knowledge to how aDNA characteristics
translate to downstream analyses.

The development of nf-core/eager in nextflow and the nf-core initiative will
also improve open-source community contributions to the pipeline. While nextflow
is written primarily in the language groovy, the nextflow DSL simplifies a lot
of concepts to a level intermediate level that bioinformaticians without Java/Groovy
experience can easily access (regardless of own programming language experience).
Furthermore, Nextflow places ubiquitous command-line interfaces, such as bash,
in a prominent position within the code, rather than custom java code and
classes. This is particularly relevant for researchers ancient DNA, who often
come from social sciences and do not have backgrounds or training in
computational biology. We hope this will motivate further bug fixes and feature
contributions to keep the updated pipeline with standard practises during a
longer life-cycle of this version of the pipeline. This will also be supported
by the active and welcoming nf-core community who provide general guidance and
advice on developing nextflow and nf-core pipelines.


## Conclusion

nf-core/eager is an efficient, portable, and accessible pipeline for processing
ancient DNA genomic data. The re-implementation with nextflow and nf-core
will improve reproducibility and inclusion of rapidly increasing ancient DNA
datasets, for both large and small laboratories. Extensive documentation also
allow newcomers to the field get a practical understanding on how to interpret
ancient DNA in the context of NGS data processing. Finally nf-core/eager brings
in the latest tools and routine screening analysis commonly used in the field,
and sets up the pipeline staying at the forefront of palaeogenetic analysis. 


## Methods

### Installation

nf-core/eager requires java, nextflow and either a environment/container system.

### Running

Further customisation can be made using profiles

Basic input is FASTQ or BAMs, FASTA reference. Complex and/or heterogeneous 
input can be supplied with a TSV file.

Runs can be monitored via terminal console or tower.nf.

Example data is included.

Output includes both files for close inspection and downstream analysis, as
well as the emailable MultiQC report.

## Competing Interests

No competing interests are declared.

## Data and software availability

All code is available on github at
[https://github.com/nf-core/eager](https://github.com/nf-core/eager) and 
archived with Zenodo under the 
DOI [10.5281/zenodo.1465061](https://doi.org/10.5281/zenodo.1465061). All
test data is from the ENA public repository.



## Acknowledgements

We thank the nf-core community for general support and suggestions during the
writing of the pipeline. We also thank Arielle Munters, Hester van Schalkwyk,
Irina Velsko, Katerine Eaton, Luc Venturini, Marcel Keller, Pierre Lindenbaum,
Pontus Skoglund, Raphael Eisenhofer, Torsten Günter, Kevin Lord for bug reports
and feature suggestions. We are grateful to the members of the Department of
Archaeogenetics at the Max Planck Institute for the Science of Human History who
performed beta testing of the pipeline. We thank the aDNA twitter community for
responding to polls regarding design decisions of the pipeine. 

We also acknowledge Christina Warinner and the Max Planck Society who provided
funds for travel to nf-core events.


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>
