---
author-meta:
- James A. Fellows Yates
- Thiseas C. Lamnidis
- Maxime Borry
- "Aida Andrades Valtue\xF1a"
- "Zandra Fagern\xE4s"
- Stephen Clayton
- Maxime U. Garcia
- Judith Neukamm
- Alexander Peltzer
bibliography:
- content/manual-references.json
date-meta: '2020-06-08'
header-includes: "<!--\nManubot generated metadata rendered from header-includes-template.html.\nSuggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html\n-->\n<meta name=\"dc.format\" content=\"text/html\" />\n<meta name=\"dc.title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"citation_title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"og:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"twitter:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"dc.date\" content=\"2020-06-08\" />\n<meta name=\"citation_publication_date\" content=\"2020-06-08\" />\n<meta name=\"dc.language\" content=\"en-GB\" />\n<meta name=\"citation_language\" content=\"en-GB\" />\n<meta name=\"dc.relation.ispartof\" content=\"Manubot\" />\n<meta name=\"dc.publisher\" content=\"Manubot\" />\n<meta name=\"citation_journal_title\" content=\"Manubot\" />\n<meta name=\"citation_technical_report_institution\" content=\"Manubot\" />\n<meta name=\"citation_author\" content=\"James A. Fellows Yates\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5585-6277\" />\n<meta name=\"twitter:creator\" content=\"@jafellowsyates\" />\n<meta name=\"citation_author\" content=\"Thiseas C. Lamnidis\" />\n<meta name=\"citation_author_institution\" content=\"Population Genetics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-4485-8570\" />\n<meta name=\"twitter:creator\" content=\"@TCLamnidis\" />\n<meta name=\"citation_author\" content=\"Maxime Borry\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-9140-7559\" />\n<meta name=\"twitter:creator\" content=\"@notmaxib\" />\n<meta name=\"citation_author\" content=\"Aida Andrades Valtue\xF1a\" />\n<meta name=\"citation_author_institution\" content=\"Computational Pathogenomics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-1737-2228\" />\n<meta name=\"twitter:creator\" content=\"@aidaanva\" />\n<meta name=\"citation_author\" content=\"Zandra Fagern\xE4s\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2667-3556\" />\n<meta name=\"twitter:creator\" content=\"@ZandraSelina\" />\n<meta name=\"citation_author\" content=\"Stephen Clayton\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5223-9695\" />\n<meta name=\"citation_author\" content=\"Maxime U. Garcia\" />\n<meta name=\"citation_author_institution\" content=\"Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2827-9261\" />\n<meta name=\"twitter:creator\" content=\"@gau\" />\n<meta name=\"citation_author\" content=\"Judith Neukamm\" />\n<meta name=\"citation_author_institution\" content=\"Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Z\xFCrich, Switzerland\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-8141-566X\" />\n<meta name=\"twitter:creator\" content=\"@JudithNeukamm\" />\n<meta name=\"citation_author\" content=\"Alexander Peltzer\" />\n<meta name=\"citation_author_institution\" content=\"Quantitative Biology Center (QBiC), Eberhard-Karls-Universit\xE4t, T\xFCbingen, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-6503-2180\" />\n<meta name=\"twitter:creator\" content=\"@alex_peltzer\" />\n<link rel=\"canonical\" href=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"og:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"twitter:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_fulltext_html_url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_pdf_url\" content=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"application/pdf\" href=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"text/html\" href=\"https://apeltzer.github.io/eager2-paper/v/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/\" />\n<meta name=\"manubot_html_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/\" />\n<meta name=\"manubot_pdf_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/manuscript.pdf\" />\n<meta property=\"og:type\" content=\"article\" />\n<meta property=\"twitter:card\" content=\"summary_large_image\" />\n<meta property=\"og:image\" content=\"https://github.com/apeltzer/eager2-paper/raw/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/content/images/nf-core-eager_social_preview.png\" />\n<meta property=\"twitter:image\" content=\"https://github.com/apeltzer/eager2-paper/raw/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/content/images/nf-core-eager_social_preview.png\" />\n<link rel=\"icon\" type=\"image/png\" sizes=\"192x192\" href=\"https://manubot.org/favicon-192x192.png\" />\n<link rel=\"mask-icon\" href=\"https://manubot.org/safari-pinned-tab.svg\" color=\"#ad1457\" />\n<meta name=\"theme-color\" content=\"#ad1457\" />\n<!-- end Manubot generated metadata -->"
keywords:
- nf-core
- nextflow
- ancientDNA
- aDNA
- palaeogenetics
- archaeogenetics
- bioinformatics
lang: en-GB
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager
...






<small><em>
This manuscript
([permalink](https://apeltzer.github.io/eager2-paper/v/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185/))
was automatically generated
from [apeltzer/eager2-paper@5eaa02c](https://github.com/apeltzer/eager2-paper/tree/5eaa02c2a94bce5d7abb60d4ea3d70b9b9700185)
on June 8, 2020.
</em></small>

## Authors



+ **James A. Fellows Yates**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5585-6277](https://orcid.org/0000-0001-5585-6277)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jfy133](https://github.com/jfy133)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [jafellowsyates](https://twitter.com/jafellowsyates)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Thiseas C. Lamnidis**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4485-8570](https://orcid.org/0000-0003-4485-8570)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [TCLamnidis](https://github.com/TCLamnidis)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [TCLamnidis](https://twitter.com/TCLamnidis)<br>
  <small>
     Population Genetics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Maxime Borry**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-9140-7559](https://orcid.org/0000-0001-9140-7559)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [maxibor](https://github.com/maxibor)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [notmaxib](https://twitter.com/notmaxib)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Aida Andrades Valtueña**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-1737-2228](https://orcid.org/0000-0002-1737-2228)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [aidaanva](https://github.com/aidaanva)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [aidaanva](https://twitter.com/aidaanva)<br>
  <small>
     Computational Pathogenomics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Zandra Fagernäs**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2667-3556](https://orcid.org/0000-0003-2667-3556)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ZandraFagernas](https://github.com/ZandraFagernas)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [ZandraSelina](https://twitter.com/ZandraSelina)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Stephen Clayton**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5223-9695](https://orcid.org/0000-0001-5223-9695)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [sc13-bioinf](https://github.com/sc13-bioinf)<br>
  <small>
     Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Maxime U. Garcia**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2827-9261](https://orcid.org/0000-0003-2827-9261)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [MaxUlysse](https://github.com/MaxUlysse)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [gau](https://twitter.com/gau)<br>
  <small>
     Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden
     · Funded by Barncancerfonden
  </small>

+ **Judith Neukamm**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8141-566X](https://orcid.org/0000-0001-8141-566X)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [JudithNeukamm](https://github.com/JudithNeukamm)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [JudithNeukamm](https://twitter.com/JudithNeukamm)<br>
  <small>
     Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Zürich, Switzerland
  </small>

+ **Alexander Peltzer**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-6503-2180](https://orcid.org/0000-0002-6503-2180)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [apeltzer](https://github.com/apeltzer)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [alex_peltzer](https://twitter.com/alex_peltzer)<br>
  <small>
     Quantitative Biology Center (QBiC), Eberhard-Karls-Universität, Tübingen, Germany
  </small>



## Abstract {.page_break_before}

The broadening utilization of ancient DNA (aDNA) to address archaeological,
palaeontological and biological questions is resulting in a rising diversity in
the size of laboratories and scale of analyses being performed. In the context
of this heterogeneous landscape, we present nf-core/eager, an advanced and
entirely redesigned pipeline for the analysis of ancient DNA genomic data.
nf-core/eager builds on existing ideas and concepts introduced in the original
EAGER pipeline, and improves various aspects of the analysis procedure by
building on computational frameworks such as Nextflow and nf-core. The pipeline
aims to address three main points: accessibility and adaptability to different
research groups and their computing configurations, reproducibility to ensure
robust analytical standards in the field, and updating the EAGER pipeline to the
latest routine ancient genomic practises. This new version of EAGER has been
developed within the nf-core initiative, to ensure high quality software
development and maintenance support; contributing to a long-term lifecycle for
the pipeline. nf-core/eager will assist in ensuring that ancient DNA sequencing
data can be utilised by a diverse range of research groups and fields.


## Introduction

Ancient DNA (aDNA) has become a widely accepted source of biological data,
helping to provide new perspective for a range of fields including archaeology,
ecology, cultural heritage, and palaeontology. The utilisation of
next-generation-sequencing has allowed the recovery of aDNA from a wide variety
of sources, including but not limited to, the skeletal remains of animals
[@doi:10.1016/j.cub.2015.04.007; @doi:10.1038/nature12323;
@doi:10.1073/pnas.1901169116; @doi:10.1073/pnas.1710186114], modern and archaic
humans [@doi:10.1038/s41586-018-0094-2; @doi:10.1126/science.1224344;
@doi:10.1038/s41586-018-0455-x], bacteria [@doi:10.1038/nature13591;
@doi:10.1073/pnas.1812865115; @doi:10.1371/journal.ppat.1006997], viruses
[@doi:10.1038/s41586-018-0097-z; @doi:10.7554/eLife.36666], plants
[@doi:10.1111/eva.12594; @doi:10.1038/s41559-019-0921-3], coprolites
[@doi:10.1016/j.chom.2019.08.018; @doi:10.7717/peerj.9001], dental calculus
[@doi:10.1038/ng.2906; @doi:10.1038/nature21674], sediments
[@doi:10.1038/nature12921; @doi:10.1126/science.aam9695], medical slides
[@doi:10.1093/molbev/msz264], parchment [@doi:10.1098/rstb.2013.0379], and most
recently ancient ‘chewing gum’ [@doi:10.1038/s41467-019-13549-9;
@doi:10.1038/s42003-019-0399-1]. Improvement in laboratory protocols to increase
yields of otherwise trace amount of DNA has at the same time led to studies that
can total hundreds of ancient individuals [@doi:10.1038/nature25738;
@doi:10.1038/nature25778], spanning single [@doi:10.1038/nature10549] to
thousands of organisms [@doi:10.1038/ng.2906]. These differences of disciplines
have led to a heterogeneous landscape in terms of types of analyses and require
different types of computing resources for each lab [@doi:10.1093/bib/bbu022;
@doi:10.3389/fgene.2018.00575]. Taking into consideration the unequal
distribution of resources (and infrastructure such as internet connection),
streamlined and efficient pipelines can help increase accessibility to
high-quality analyses

The degraded nature of aDNA poses an extra layer of complexity to standard
modern genomic analysis. Through a variety of processes [@doi:10.1038/362709a0],
DNA molecules will fragment overtime resulting in ultra-short molecules
[@doi:10.1038/nature17405]). These sequences have low nucleotide complexity
making it difficult to identify with precision which part of the genome a read
is derived from. Furthermore, when fragmentation is not a 'clean break' this can
lead to uneven ends with single-stranded 'overhangs' at end of molecules, which
are susceptible to chemical processes such as deamination that lead to
mis-incorporation of bases during library construction
[@doi:10.1073/pnas.0704665104]. On top of this, taphonomic processes such as
heat, moisture, and microbial and burial environment processes lead to varying
rates of degradation [@doi:10.1093/nar/gkx361;
@doi:10.1146/annurev-genom-091416-035526], where the original DNA content of a
sample is lost and supplanted by modern environmental DNA. Later handling by
archaeologists, museum curators, and scientists can also contribute 'modern'
contamination. While these characteristics can help provide evidence towards the
'authenticity' of true aDNA sequences (e.g. aDNA C>T 'damage' profiles
[@doi:10.1093/bioinformatics/btt193]), they also pose specific challenges such
as unspecific DNA alignment and/or low coverage and miscoding lesions resulting
in low-confidence genotyping. These factors often lead to prohibitive sequencing
costs to retrieve enough data for modern NGS data pipelines (e.g. > 1 billion
reads for a 1X depth coverage _Yersinia pestis_ genome
[@doi:10.1016/j.cell.2015.10.009]), and thus require aDNA tailored methods and
techniques to overcome these challenges.

Two previously published and commonly used pipelines in the field are PALEOMIX
[@doi:10.1038/nprot.2014.063] and EAGER [@doi:10.1186/s13059-016-0918-z]. These
two pipelines take a similar approach to link together standard tools used for
Illumina NGS data processing (sequencing quality control, sequencing adapter
removal/and or paired-end read merging, mapping of reads to a reference genome,
genotyping, etc.), but with a specific focus on tools that are designed for or
well-suited for aDNA (such as bwa aln that works well on ultra-short molecules
[@doi:10.1093/bioinformatics/btp324] and mapDamage
[@doi:10.1093/bioinformatics/btr347] for aDNA characteristics evaluation). Yet,
neither of these pipelines have had major updates to bring them in-line with
current routine aDNA analyses. Metagenomic screening of off-target genomic reads
for pathogens or microbiomes [@doi:10.1038/ng.2906; @doi:10.1038/nature21674]
has become particularly common, given its role in revealing widespread
infectious disease and possible epidemics that had previously been undetected in
the archaeological record [@doi:10.1016/j.cell.2015.10.009;
@doi:10.1016/j.cub.2017.10.025; @doi:10.1038/s41586-018-0097-z;
@doi:10.7554/eLife.36666]. Without easy access to the latest field-established
analytical routines, aDNA studies risk being published without the necessary
quality control checks that ensure aDNA authenticity and without yielding the
full range of possibilities from their data.

To address these shortcomings, we have completely re-implemented the latest
version of the EAGER pipeline in Nextflow [@doi:10.1038/nbt.3820] (a
domain-specific-language (DSL), specifically designed for the construction of
omics analysis pipelines), introduced new features, and more flexible pipeline
configurations. In addition, the newly named pipeline - nf-core/eager - has been
developed in the context of the nf-core community framework
[@doi:10.1038/s41587-020-0439-x], which enforces strict guidelines for
best-practises in software development.


# Results and Discussion

## Scalability, Portability, and Efficiency

The reimplementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager into
various job schedulers in POSIX High-Performance-Cluster (HPC) environments,
cloud computing resources, and as well as local workstations. This portability
allows both small and big labs to run nf-core/eager regardless of the type of
computer or cluster size, with minimal effort or configuration, facilitating
reproducibility and therefore maintenance of standards within the field. This is
further assisted by the in-built compatibility with software environments and
containers such as Conda [@url:https://conda.io],
Docker[@url:https://www.docker.com] and Singularity [@url:https://sylabs.io].
These are single packages that include all the software (with exact versions)
required by the pipeline, in a form that is installable and runnable regardless
of the setup of their local software environment. Another major change with
nf-core/eager is that the graphical-user-interface (GUI) set up of an
nf-core/eager run is now replaced with a command-line-interface (CLI) as the
primary user interaction mode. This is more compatible and portable with most
HPCs (that may not offer display of program windows), and is in line with the
vast majority of bioinformatic tools. We therefore believe this will not be a
hindrance to new researchers from outside computational biology. However there
are plans from Nextflow (with tower.nf [@url:https://tower.nf]) and within the
nf-core community to provide multiple alternatives in the near future including
a CLI wizard and a web-based input GUI.

Secondly, reproducibility is made easier through the use of 'profiles' that can
define configuration parameters. These profiles can be managed at different
hierarchical levels. HPC-level profiles can specify parameters for the computing
environment (job schedulers, cache locations for containers, maximum memory and
CPU resources etc.), which can be centrally managed to ensure all users of a
group use the same settings. Pipeline-level profiles, specifying default
parameters for nf-core/eager itself, allow fast access to routine pipeline-run
parameters via a single flag in the nf-core/eager run command, without having to
configure each new run from scratch.

Compared to the original EAGER that utilised per-FASTQ XML files with hardcoded
filepaths for a specific user's server, nf-core/eager allows researchers to
publish the specific profile used in their runs alongside their publications, to
ensure other groups can generate the same results. Usage of profiles also
reduces mistakes caused by insufficient 'prose' based reporting of program
settings that can be regularly found in the literature. The default
nf-core/eager profile uses parameters evaluated in different aDNA specific
contexts (e.g. in @doi:10.1186/1471-2164-13-178], and will be updated in each
new release as new studies are published.

nf-core/eager provides improved efficiency over the original EAGER pipeline by
replacing the sample-by-sample sequential processing with Nextflow's
asynchronous parallelisation, whereby multiple pipeline steps and samples are
run in parallel (in addition to single pipeline step multi-threading). This,
combined with pre-defined per-process customisation of resource parameters,
reduces unnecessary resource allocation that can occur with new users to each
step of an NGS data processing pipeline. This is particularly pertinent given
the increasing use of centralised HPCs or cloud computing, which often use
per-hour cost calculations.

## Updated Workflow

nf-core/eager follows a similar structural foundation to the original EAGER.
Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file,
this can be split into four main stages:

1. Preprocessing
   - Sequencing quality control: FastQC
     [@url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc]
   - Sequencing artefact clean-up (merging, adapter clipping): AdapterRemoval2
     [@doi:10.1186/s13104-016-1900-2]
   - Preprocessing statistics generation
2. Mapping and post- processing
   - Alignment against reference genome: BWA
     [@doi:10.1093/bioinformatics/btp324; @doi:10.1093/bioinformatics/btp698,
     @arxiv:1303.3997], CircularMapper [@doi:10.1186/s13059-016-0918-z]
   - Mapping quality filtering: Samtools [@doi:10.1093/bioinformatics/btp352]
   - PCR duplicate removal: DeDup [@doi:10.1186/s13059-016-0918-z], Picard
     MarkDuplicates [@doi:10.1101/gr.107524.110]
   - Mapping statistics generation: PreSeq [@doi:10.1038/nmeth.2375], Qualimap2
     [@doi:10.1093/bioinformatics/btv566]
3. aDNA Evaluation and Modification
   - Damage profiling: DamageProfiler [@doi:10.5281/zenodo.3557708]
   - aDNA reads selection: PMDTools [@doi:10.1073/pnas.1318934111]
   - Damage removal: Bamutils[@doi:10.1101/gr.176552.114]
   - (Human) contamination estimation: ANGSD [@doi:10.1186/s12859-014-0356-4]
4. Genotyping and Consensus Sequencing: GATK [@doi:10.1101/gr.107524.110],
   VCF2Genome [@doi:10.1186/s13059-016-0918-z]

In nf-core/eager, all tools also originally used in EAGER have been updated to
latest versions, as available on Bioconda [@doi:10.1038/s41592-018-0046-7] and
Conda-forge [@url:https://conda-forge.github.io] to ensure widespread
accessibility and stability of utilized tools. The MapDamage2 (for damage
profile generation) [@doi:10.1093/bioinformatics/btt193] and Schmutzi for
(mitochondrial contamination estimation) [@doi:10.1186/s13059-015-0776-0]
methods have not been carried over to nf-core/eager, the first because a more
performant successor method is now available (DamageProfiler), and the latter
because a stable release of the method could not be migrated to Bioconda. We
anticipate that there will be an updated version of Schmutzi in the near future
that will allow us to integrate the method again in nf-core/eager, once a
version is released on Bioconda. Support for the Bowtie2 aligner
[@doi:10.1038/nmeth.1923] will be added in the near future, after consultation
with the palaeogenetics community. New tools to the basic workflow include fastp
[@doi:10.1093/bioinformatics/bty560] for the removal of 'poly-G' sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms,
[@url:https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/]).
For genotyping we have now included FreeBayes [@arxiv:1207.3907] as an
alternative to the human-focused GATK tools. We have also maintained the
possibility of using the now unsupported GATK UnifiedGenotyper, as the GATK
HaplotypeCaller performs _de novo_ assembly around possible variants, which may
not be suitable for low-coverage aDNA data.

![Simplified schematic of the nf-core/eager workflow pipeline. Green filled
bubbles indicate new functionality added over the original EAGER
pipeline.](images/Fig1_eager2_workflow.png){#fig:workflow-schematic
width="100%"}

We have further extended the genomic analysis functionality of the pipeline by
adding ancient metagenomic analysis, to identify the wider taxonomic content of
a sample. We have added the ability to screen all off-target reads (not mapped
to the reference genome) with two metagenomic profilers: MALT
[@doi:10.1101/050559; @doi:10.1038/s41559-017-0446-6] and Kraken2
[@doi:10.1186/s13059-019-1891-0]. Characterisation of properties of authentic
aDNA from MALT alignments is carried out with the HOPS pipeline
[@doi:10.1186/s13059-019-1903-0]. Ancient metagenomic studies sometimes may
include comparative samples from living individuals
[@doi:10.1186/s40168-019-0717-3)]. To support open data, whilst respecting data
privacy, nf-core/eager includes a 'strip_fastq' script which creates raw FASTQ
files, but with reference-genome mapped reads removed. This allows safe upload
of sequencing data to public repositories with identifiable human data removed.

Additional functionality tailored for ancient bacterial genomics includes
integration of a SNP alignment generation tool, MultiVCFAnalyzer
[@doi:10.1038/nature13591], which allows assessment of cross-mapping levels from
different related taxa to a reference genome - a common challenge in ancient
bacterial genome reconstruction [@doi:10.1146/annurev-genom-091416-035526]. The
output SNP alignment FASTA file can then be used for downstream analyses such as
phylogenetic tree construction. Simple coverage statistics of particular
annotations (e.g. genes) of an input reference is offered by bedtools
[@doi:10.1093/bioinformatics/btq033], which can be used, for example, for
determining functional differences between ancient bacterial strains (as in
[@doi:10.1016/j.cub.2017.10.025]). When using a human reference genome,
nf-core/eager can now also can give estimates of the biological sex of a given
individual with Sex.DetEERRmine [@doi:10.1038/s41467-018-07483-5]. A dedicated
'endogenous DNA' calculator (endorS.py) is also included to provide a percentage
estimate of the sequenced reads matching the reference from the total number of
reads sequenced per library.

Given the large amount of sequencing often required to yield sufficient genome
coverage from aDNA data, palaeogeneticists tend to use multiple (differently
treated) libraries or sequencing runs. The original EAGER pipeline could only
run single libraries at a time, and in these contexts required significant
manual user input in merging different FASTQ or BAM files. A major upgrade in
nf-core/eager is that the new pipeline supports automated processing of complex
sequencing strategies for many samples. As an alternative to direct paths to
FASTQ or BAM files, the pipeline can also accepts a simple table in TSV format
which includes file paths and additional metadata such as sample name, library
name, sequencing lane, colour chemistry, and UDG treatment. This allows
simultaneous processing and appropriate merging of heterogeneous data from
multiple sequencing runs and/or library types.

The original EAGER pipeline required users to look through many independent
output directories and files to make full assessment of their sequencing data.
This has now been replaced with a much more extensive MultiQC
[@doi:10.1093/bioinformatics/btw354] report. This tool aggregates the log files
of every supported tool into a single interactive report, and assists users in
making fuller assessment of their sequencing and analysis runs. Most tools
within nf-core/eager have a corresponding MultiQC module to enable comprehensive
evaluation of all stages of the pipeline.

An overview of the entire pipeline is shown in Figure {@fig:workflow-schematic}.

## Accessibility

Alongside the interactive MultiQC report, we have written extensive
documentation on all parts of running and interpreting the output of the
pipeline. Given that a large fraction of aDNA researchers come from fields
outside computational biology, and thus may have limited computational training,
we have written documentation that also gives guidance on how to interpret each
section of the report, specifically in the context of NGS and aDNA. This
includes schematic images of best practices or expected output that are
published under CC-BY licenses to allow for use in other training material
(example in {@fig:doc-image-example}). We hope this open-access resource will
make the study of aDNA more accessible to researchers new to the field, by
providing practical guidelines how to evaluate characteristics and effects of
aDNA on downstream analyses.

![Example of output interpretation documentation schematic images that can
assist new users in the interpretation to next-generation-sequencing aDNA
processing.](images/Fig2_fastqc_adapter_content.png){#fig:doc-image-example
width="70%"}

The development of nf-core/eager in Nextflow and the nf-core initiative will
also improve open-source community contributions to the pipeline. While Nextflow
is written primarily in Groovy, the Nextflow DSL simplifies a number of concepts
to an intermediate level that bioinformaticians without Java/Groovy experience
can easily access (regardless of own programming language experience).
Furthermore, Nextflow places ubiquitous and more widely known command-line
interfaces, such as bash, in a prominent position within the code, rather than
custom java code and classes. We hope this will motivate further bug fixes and
feature contributions from the community, to keep the pipeline state-of-the-art
and ensure a longer life-cycle. This will also be supported by the active and
welcoming nf-core community who provide general guidance and advice on
developing Nextflow and nf-core pipelines.


## Conclusion

nf-core/eager is an efficient, portable, and accessible pipeline for processing
aDNA genomic data. This re-implementation of EAGER into Nextflow and nf-core
will improve reproducibility and inclusion of rapidly increasing aDNA datasets,
for both large and small laboratories. Extensive documentation also enables
newcomers to the field get a practical understanding on how to interpret aDNA in
the context of NGS data processing. Ultimately, nf-core/eager provides easier
access to the latest tools and routine screening analyses commonly used in the
field, and sets up the pipeline for staying at the forefront of palaeogenetic
analysis.


## Methods

### Installation

nf-core/eager requires a version of Java, Nextflow and either a functional Conda
installation *or* Docker/Singularity container installation. A quick
installation guide to follow to get started can be found in the *Quickstart*
section of the nf-core/eager repository
[@url:https://github.com/nf-core/eager#quick-start].

### Running

After the installation, users can run the pipeline using standard test data by
utilizing some of the `test` profiles we provide (e.g. using Docker):

```bash
nextflow run nf-core/eager -r 2.1.0 -profile test,docker
```

This will download test data automatically, run the pipeline locally with all
software tools containerized in a Docker image and store the output of that run
in the `./results` folder of your current directory.

The default pipeline settings assumes paired end FASTQ data and will run:
FastQC; AdapterRemoval2 (merging and adapter clipping); post-clipping FastQC
(for AdapterRemoval2 performance evaluation); bwa mapping (with the 'aln'
algorithm); samtools flagstat (for mapping statistics); endorS.py (for
endogenous DNA calculation); DeDup (for PCR amplicon deduplication); PreSeq (for
library complexity evaluation); DamageProfiler and Qualimap2 (for genome
coverage statistics) and the MultiQC pipeline run report. If no additional FASTA
indices are given, these will also be generated.

The pipeline is highly configurable and most modules can be turned on and off at
the request of the user using different flags to allow high customisation to
each users needs. For example, to include metagenomic screening of off-target
reads and sex determination based on on-target mappings of pre-clipped
single-end data:

```bash
nextflow run nf-core/eager -r 2.1.0 -profile conda --input '/<path>/<to>/*/*R1*.fastq.gz' --single_end --fasta '/<path>/<to>/<reference>.fasta.gz' --skip_fastqc --skip_adapterremoval --run_bam_filtering --bam_discard_unmapped --bam_unmapped_type 'fastq' --run_metagenomic_screening --metagenomic_tool 'malt' --database '/<path>/<to>/<malt_database>' --run_sexdeterrmine
```

#### Profiles

We utilize a central configuration profile repository to enable users from
various institutions to use pipelines on their particular infrastructure more
easily [@url:https://github.com/nf-core/configs]. There are multiple resources
listed in this repository with information on how to add your own configuration
profile with help from the nf-core community.

Users can customize this infrastructure profile by themselves, with the nf-core
community, or with their local system administrator to make sure that the
pipeline runs successfully, and can then rely on the Nextflow and nf-core
framework to ensure compatibility upon further infrastructure changes. For
example, in order to run the nf-core/eager pipeline at the Max Planck Institute
for the Science of Human History (MPI-SHH) in Jena, users only have to run:

```bash
nextflow run nf-core/eager -r 2.1.0 -profile shh_cdag --input '/<path>/<to>/*/*{R1,R2}*.fastq.gz' --fasta '/<path>/<to>/<reference>.fasta.gz'
```

This runs the testing profile of the nf-core/eager pipeline with parameters
specifically adapted to the HPC system at the MPI-SHH. In some cases, similar
institutional configs for other institutions may already exist (originally
utilised for different nf-core pipelines), so users need not write their own.

#### Inputs

The pipeline can be started using (raw) FASTQ files from sequencing or
pre-mapped BAM files. Additionally, the pipeline requires a FASTA reference
genome.

If BAM input is provided, an optional conversion to FASTQ is offered, otherwise
BAM files processing will start from the post-mapping stage.

If users have complex set-ups, e.g. multiple sequencing lanes that require
merging of files for example, the pipeline can be supplied with a tabular
separated value (TSV) file to enable such complex data handling. Both FASTQs and
BAMs can be provided in this set up. FASTQs with the same library name and
sequencing chemistry but sequenced across multiple lanes will be concatenated
after AdapterRemoval and prior mapping. Libraries with the sample name and with
the same UDG treatment, will be merged after deduplication. If libraries with
the sample name have different UDG treatment, these will be merged after the
aDNA modification stage (i.e. BAM trimming or PMDtools, if turned on), prior
genotyping, as shown in in Figure {@fig:merging-files}.

![Schematic of different processing and merging points based on the nature of
different libraries, as specified in metadata of a TSV file. Dashed boxes
represent optional library-specific
processes](images/Fig3_merging_files.png){#fig:merging-files width="70%"}

As Nextflow will automatically download files from URLs, profiles and/or TSV
files can include links to publicly available data (e.g. the ENA FTP server).
This assists in reproducibility as if profiles or TSV files are uploaded with a
publication, a researcher wishing to re-analyse the data in the same way can use
the exact settings and merging procedures in the original publication, without
having to reconstruct this from prose.

#### Monitoring

Users can either monitor their pipeline execution with the messages Nextflow
prints to the console while running, or utilize projects such as Nextflow Tower
[@url:https://tower.nf] to monitor their analysis pipeline during runtime.

#### Output

The pipeline produces several dozen output files in various file formats, with a
more detailed listing available in the user documentation. This includes
metrics, statistical analysis data and standardized output files (BAM, VCF) for
close inspection and further downstream analysis, as well as a MultiQC report.
If an emailing daemon is set up on the server, the latter can even be emailed to
users automatically, when starting the pipeline with a dedicated option (--email
you@yourdomain.org).

## Data and software availability

All code is available on github at
[https://github.com/nf-core/eager](https://github.com/nf-core/eager) and
archived with Zenodo under the DOI
[10.5281/zenodo.1465061](https://doi.org/10.5281/zenodo.1465061). The version of
nf-core/eager that this preprint is based on the current dev branch of the
GitHub repository (2.2.0dev), and on publication will be released as 2.2.0.

This paper was written with Manubot [@doi:10.1371/journal.pcbi.1007128].


## Competing Interests

No competing interests are declared.


## Acknowledgements

We thank the nf-core community for general support and suggestions during the
writing of the pipeline. We also thank Arielle Munters, Hester van Schalkwyk,
Irina Velsko, Katerine Eaton, Luc Venturini, Marcel Keller, Pierre Lindenbaum,
Pontus Skoglund, Raphael Eisenhofer, Torsten Günter, and Kevin Lord for bug
reports and feature suggestions. We are grateful to the members of the
Department of Archaeogenetics at the Max Planck Institute for the Science of
Human History who performed beta testing of the pipeline. We thank the aDNA
twitter community for responding to polls regarding design decisions of the
pipeline.

We also want to thank Selina Carlhoff, Alexander Herbig and Wolfgang Haak for
kindly providing comments on this manuscript.

We also want to acknowledge Christina Warinner, Stephan Schiffels and the Max
Planck Society who provided funds for travel to nf-core events.


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>
