## Introduction

Ancient DNA (aDNA) has become a widely accepted source of biological data, helping to provide new perspective for a range of fields including archaeology, ecology, cultural heritage, and palaeontology. The utilisation of next-generation-sequencing has allowed the recovery of aDNA from a wide variety of sources, including but not limited to, the skeletal remains of animals [@doi:10.1016/j.cub.2015.04.007, @doi:10.1038/nature12323, @doi:10.1073/pnas.1901169116, @doi:10.1073/pnas.1710186114], modern and archaic humans [@doi:10.1038/s41586-018-0094-2, @doi:10.1126/science.1224344, @doi:10.1038/s41586-018-0455-x], bacteria [@doi:10.1038/nature13591, @doi:10.1073/pnas.1812865115, @doi:10.1371/journal.ppat.1006997], viruses [@doi:10.1038/s41586-018-0097-z, @doi:10.7554/eLife.36666], plants [@doi:10.1111/eva.12594, @doi:10.1038/s41559-019-0921-3], coprolites [@doi:10.1016/j.chom.2019.08.018, @doi:10.7717/peerj.9001], dental calculus [@doi:10.1038/ng.2906, @doi:10.1038/nature21674], sediments [@doi:10.1038/nature12921, @doi:10.1126/science.aam9695], medical slides [@doi:10.1093/molbev/msz264], parchment [@doi:10.1098/rstb.2013.0379], and most recently ancient ‘chewing gum’ [@doi:10.1038/s41467-019-13549-9, @doi:10.1038/s42003-019-0399-1]. Improvement in laboratory protocols to increase yields of otherwise trace amount of DNA has at the same time led to studies that can total hundreds of ancient individuals [@doi:10.1038/nature25738, @doi:10.1038/nature25778], spanning single [@doi:10.1038/nature10549] to thousands of organisms [@doi:10.1038/ng.2906]. These differences of disciplines have led to a heterogeneous landscape in terms of types of analyses, and thus various types of computing resources required by different labs. Particularly, this is true for regions of the world where aDNA as a field of research is still emerging.

The degraded nature of aDNA poses an extra layer of complexity to standard modern genomic analysis. Through a variety of processes [@doi:10.1038/362709a0], DNA molecules will fragment overtime resulting in ultra-short molecules (e.g. down to 35 bp [@doi:10.1038/nature17405]). These sequences low nucleotide complexity making it difficult identify with precision which part of the genome a read is derived from. Furthermore, when fragmentation is not a 'clean break' this can lead to uneven ends with single-stranded 'overhangs' at end of molecules, which are susceptible to chemical processes such as deamination that are misread by some polymerases [@doi:10.1073/pnas.0704665104], resulting in incorrect base calling during sequencing. On top of this, taphonomic processes such as heat, moisture, and microbial and burial environment processes leads to varying rates of degradation [@doi:10.1093/nar/gkx361, @doi:10.1146/annurev-genom-091416-035526], where the original DNA content of a sample is loss and supplanted by modern environmental DNA. Later handling by modern archaeologists, museum curators, and scientists can also contribute 'modern' contamination. While these characteristics can help provide evidence towards the 'authenticity' of true aDNA sequences (e.g. aDNA C>T 'damage' profiles [@doi:10.1093/bioinformatics/btt193]), they also pose specific challenges such as unspecific DNA alignment and/or low coverage and miscoding lesions resulting in low-confidence genotyping. To retrieve enough data for modern NGS data pipelines these factors often lead to prohibitive sequencing costs (e.g. > 1 billion reads for a 1X depth coverage _Y. pestis_ genome [@doi:10.1016/j.cell.2015.10.009]), and thus require aDNA requires tailored methods and techniques to overcome these challenges.

Two previously published and commonly used pipelines in the field are PALEOMIX [@doi:10.1038/nprot.2014.063] and EAGER [@doi:10.1186/s13059-016-0918-z]. These two pipelines take a similar approach to link together standard tools used for Illumina NGS data processing (sequencing quality control, sequencing adapter removal/and or paired-end read merging, mapping of reads to a reference genome, genotyping, etc.), but with a specific focus on tools that are designed for or well-suited for aDNA (such as bwa aln that works well on ultra-short molecules [@doi:10.1093/bioinformatics/btp324] and mapDamage [@doi:10.1093/bioinformatics/btr347] for aDNA characteristics evaluation). Yet, neither of these pipelines have had major updates to bring them in-line with current routine aDNA analyses. Metagenomic screening of off-target genomic reads for pathogens or microbiomes [@doi:10.1038/ng.2906, @doi:10.1038/nature21674] has become particularly common, given it's role in revealing widespread infectious disease and possible epidemics that had previously been undetected in the archaeological record [@doi:10.1016/j.cell.2015.10.009, @doi:10.1016/j.cub.2017.10.025, @doi:10.1038/s41586-018-0097-z, @doi:10.7554/eLife.36666]. Without easy access to the latest field-established analytical routines, aDNA studies from groups new to the field risk being published without the necessary quality control checks that ensure aDNA authenticity and without yielding the full range of possibilities from their data.

To address these shortcomings, we have completely re-implemented the latest version of the EAGER pipeline in Nextflow [@doi:10.1038/nbt.3820] (a domain-specific-language (DSL), specifically designed for the construction of omics analysis pipelines), introduced new features, and more flexible pipeline configurations. In addition, the newly named pipeline - nf-core/eager - has been developed in the context of the nf-core community framework [@doi:10.1038/s41587-020-0439-x], which enforces strict guidelines for best-practises in software development.
