# Results and Discussion

## Scalability, Portability, and Efficiency

The reimplementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager into
various job schedulers in POSIX High-Performance-Cluster (HPC) environments,
cloud computing resources, as well as local workstations. This portability
allows both small and big labs to run nf-core/eager regardless of the type of
computer or cluster size, with minimal effort or configuration, facilitating
reproducibility and therefore maintenance of standards within the field. This is
further assisted by the in-built compatibility with software environments and
containers such as Conda [@url:https://conda.io],
Docker[@url:https://www.docker.com] and Singularity [@url:https://sylabs.io].
These are single packages that include all the software (with exact versions)
required by the pipeline, in a form that is installable and runnable regardless
of the setup of their local software environment. Another major change with
nf-core/eager is that the graphical-user-interface (GUI) set up of an
nf-core/eager run is now replaced with a command-line-interface (CLI) as the
primary user interaction mode. This is more compatible and portable with most
HPCs (that may not offer display of program windows), and is in line with the
vast majority of bioinformatic tools. We therefore believe this will not be a
hindrance to new researchers from outside computational biology. However there
are plans from Nextflow (with tower.nf [@url:https://tower.nf]) and within the
nf-core community to provide multiple alternatives in the near future including
a CLI wizard and a web-based input GUI.

Secondly, reproducibility is made easier through the use of 'profiles' that can
define configuration parameters. These profiles can be managed at different
hierarchical levels. HPC-level profiles can specify parameters for the computing
environment (job schedulers, cache locations for containers, maximum memory and
CPU resources etc.), which can be centrally managed to ensure all users of a
group use the same settings. Pipeline-level profiles, specifying default
parameters for nf-core/eager itself, allow fast access to routine pipeline-run
parameters via a single flag in the nf-core/eager run command, without having to
configure each new run from scratch.

Compared to the original EAGER that utilised per-FASTQ XML files with hardcoded
filepaths for a specific user's server, nf-core/eager allows researchers to
publish the specific profile used in their runs alongside their publications, to
ensure other groups can generate the same results. Usage of profiles also
reduces mistakes caused by insufficient 'prose' based reporting of program
settings that can be regularly found in the literature. The default
nf-core/eager profile uses parameters evaluated in different aDNA specific
contexts (e.g. in @doi:10.1186/1471-2164-13-178], and will be updated in each
new release as new studies are published.

nf-core/eager provides improved efficiency over the original EAGER pipeline by
replacing the sample-by-sample sequential processing with Nextflow's
asynchronous parallelisation, whereby multiple pipeline steps and samples are
run in parallel (in addition to single pipeline step multi-threading). This,
combined with pre-defined per-process customisation of resource parameters,
reduces unnecessary resource allocation that can occur with new users to each
step of an NGS data processing pipeline. This is particularly pertinent given
the increasing use of centralised HPCs or cloud computing, which often use
per-hour cost calculations.

## Updated Workflow

nf-core/eager follows a similar structural foundation to the original EAGER.
Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file,
this can be split into four main stages:

1. Preprocessing
   - Sequencing quality control: FastQC
     [@url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc]
   - Sequencing artefact clean-up (merging, adapter clipping): AdapterRemoval2
     [@doi:10.1186/s13104-016-1900-2]
   - Preprocessing statistics generation
2. Mapping and post- processing
   - Alignment against reference genome: BWA
     [@doi:10.1093/bioinformatics/btp324; @doi:10.1093/bioinformatics/btp698,
     @arxiv:1303.3997], CircularMapper [@doi:10.1186/s13059-016-0918-z]
   - Mapping quality filtering: Samtools [@doi:10.1093/bioinformatics/btp352]
   - PCR duplicate removal: DeDup [@doi:10.1186/s13059-016-0918-z], Picard
     MarkDuplicates [@doi:10.1101/gr.107524.110]
   - Mapping statistics generation: PreSeq [@doi:10.1038/nmeth.2375], Qualimap2
     [@doi:10.1093/bioinformatics/btv566]
3. aDNA Evaluation and Modification
   - Damage profiling: DamageProfiler [@doi:10.5281/zenodo.3557708]
   - aDNA reads selection: PMDTools [@doi:10.1073/pnas.1318934111]
   - Damage removal: Bamutils[@doi:10.1101/gr.176552.114]
   - (Human) contamination estimation: ANGSD [@doi:10.1186/s12859-014-0356-4]
4. Genotyping and Consensus Sequencing: GATK [@doi:10.1101/gr.107524.110],
   VCF2Genome [@doi:10.1186/s13059-016-0918-z]

In nf-core/eager, all tools also originally used in EAGER have been updated to
latest versions, as available on Bioconda [@doi:10.1038/s41592-018-0046-7] and
Conda-forge [@url:https://conda-forge.github.io] to ensure widespread
accessibility and stability of utilized tools. The MapDamage2 (for damage
profile generation) [@doi:10.1093/bioinformatics/btt193] and Schmutzi for
(mitochondrial contamination estimation) [@doi:10.1186/s13059-015-0776-0]
methods have not been carried over to nf-core/eager, the first because a more
performant successor method is now available (DamageProfiler), and the latter
because a stable release of the method could not be migrated to Bioconda. We
anticipate that there will be an updated version of Schmutzi in the near future
that will allow us to integrate the method again in nf-core/eager, once a
version is released on Bioconda. Support for the Bowtie2 aligner
[@doi:10.1038/nmeth.1923] will be added in the near future, after consultation
with the palaeogenetics community. New tools to the basic workflow include fastp
[@doi:10.1093/bioinformatics/bty560] for the removal of 'poly-G' sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms,
[@url:https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/]).
For genotyping we have now included FreeBayes [@arxiv:1207.3907] as an
alternative to the human-focused GATK tools. We have also maintained the
possibility of using the now unsupported GATK UnifiedGenotyper, as the GATK
HaplotypeCaller performs _de novo_ assembly around possible variants, which may
not be suitable for low-coverage aDNA data.

![Simplified schematic of the nf-core/eager workflow pipeline. Green filled
bubbles indicate new functionality added over the original EAGER
pipeline.](images/Fig1_eager2_workflow.png){#fig:workflow-schematic
width="100%"}

We have further extended the genomic analysis functionality of the pipeline by
adding ancient metagenomic analysis, to identify the wider taxonomic content of
a sample. We have added the ability to screen all off-target reads (not mapped
to the reference genome) with two metagenomic profilers: MALT
[@doi:10.1101/050559; @doi:10.1038/s41559-017-0446-6] and Kraken2
[@doi:10.1186/s13059-019-1891-0]. Characterisation of properties of authentic
aDNA from MALT alignments is carried out with the HOPS pipeline
[@doi:10.1186/s13059-019-1903-0]. Ancient metagenomic studies sometimes may
include comparative samples from living individuals
[@doi:10.1186/s40168-019-0717-3)]. To support open data, whilst respecting data
privacy, nf-core/eager includes a 'strip_fastq' script which creates raw FASTQ
files, but with reference-genome mapped reads removed. This allows safe upload
of sequencing data to public repositories with identifiable human data removed.

Additional functionality tailored for ancient bacterial genomics includes
integration of a SNP alignment generation tool, MultiVCFAnalyzer
[@doi:10.1038/nature13591], which allows assessment of cross-mapping levels from
different related taxa to a reference genome - a common challenge in ancient
bacterial genome reconstruction [@doi:10.1146/annurev-genom-091416-035526]. The
output SNP alignment FASTA file can then be used for downstream analyses such as
phylogenetic tree construction. Simple coverage statistics of particular
annotations (e.g. genes) of an input reference is offered by bedtools
[@doi:10.1093/bioinformatics/btq033], which can be used, for example, for
determining functional differences between ancient bacterial strains (as in
[@doi:10.1016/j.cub.2017.10.025]). When using a human reference genome,
nf-core/eager can now also can give estimates of the biological sex of a given
individual with Sex.DetEERRmine [@doi:10.1038/s41467-018-07483-5]. A dedicated
'endogenous DNA' calculator (endorS.py) is also included to provide a percentage
estimate of the sequenced reads matching the reference from the total number of
reads sequenced per library.

Given the large amount of sequencing often required to yield sufficient genome
coverage from aDNA data, palaeogeneticists tend to use multiple (differently
treated) libraries or sequencing runs. The original EAGER pipeline could only
run single libraries at a time, and in these contexts required significant
manual user input in merging different FASTQ or BAM files. A major upgrade in
nf-core/eager is that the new pipeline supports automated processing of complex
sequencing strategies for many samples. As an alternative to direct paths to
FASTQ or BAM files, the pipeline can also accepts a simple table in TSV format
which includes file paths and additional metadata such as sample name, library
name, sequencing lane, colour chemistry, and UDG treatment. This allows
simultaneous processing and appropriate merging of heterogeneous data from
multiple sequencing runs and/or library types.

The original EAGER pipeline required users to look through many independent
output directories and files to make full assessment of their sequencing data.
This has now been replaced with a much more extensive MultiQC
[@doi:10.1093/bioinformatics/btw354] report. This tool aggregates the log files
of every supported tool into a single interactive report, and assists users in
making fuller assessment of their sequencing and analysis runs. Most tools
within nf-core/eager have a corresponding MultiQC module to enable comprehensive
evaluation of all stages of the pipeline.

An overview of the entire pipeline is shown in Figure {@fig:workflow-schematic}.

## Accessibility

Alongside the interactive MultiQC report, we have written extensive
documentation on all parts of running and interpreting the output of the
pipeline. Given that a large fraction of aDNA researchers come from fields
outside computational biology, and thus may have limited computational training,
we have written documentation that also gives guidance on how to interpret each
section of the report, specifically in the context of NGS and aDNA. This
includes schematic images of best practices or expected output that are
published under CC-BY licenses to allow for use in other training material
(example in {@fig:doc-image-example}). We hope this open-access resource will
make the study of aDNA more accessible to researchers new to the field, by
providing practical guidelines how to evaluate characteristics and effects of
aDNA on downstream analyses.

![Example of output interpretation documentation schematic images that can
assist new users in the interpretation to next-generation-sequencing aDNA
processing.](images/Fig2_fastqc_adapter_content.png){#fig:doc-image-example
width="70%"}

The development of nf-core/eager in Nextflow and the nf-core initiative will
also improve open-source community contributions to the pipeline. While Nextflow
is written primarily in Groovy, the Nextflow DSL simplifies a number of concepts
to an intermediate level that bioinformaticians without Java/Groovy experience
can easily access (regardless of own programming language experience).
Furthermore, Nextflow places ubiquitous and more widely known command-line
interfaces, such as bash, in a prominent position within the code, rather than
custom java code and classes. We hope this will motivate further bug fixes and
feature contributions from the community, to keep the pipeline state-of-the-art
and ensure a longer life-cycle. This will also be supported by the active and
welcoming nf-core community who provide general guidance and advice on
developing Nextflow and nf-core pipelines.
