# Results and Discussion

## Scalability, Portability, and Efficiency

The reimplementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager
into various schedulers in POSIX High-Performance-Cluster (HPC) environments, cloud computing resources and as
well as local workstations. This portability allows both small and big labs to run
nf-core/eager regardless of the type of computer or cluster cluster sizes with
minimal effort or configuration, facilitating reproducibility and therefore
maintenance of standards within the field. This is further assisted by the
in-built compatibility with software environments and containers such as conda,
docker and singularity. This ensures exact versions of software are used by
a user, regardless of the setup of their respective cluster environment. Another major
change with nf-core/eager is that the GUI input is now replaced with a
command-line-interface as the primary user interaction mode. This is more
compatible and portable with most HPCs (that may not offer X11 forwarding) and
is in line with the vast majority of bioinformatic tools. We therefore believe
this will not be a hindrance to new researchers from outside computational
biology, however there are plans within the nf-core to provide multiple
alternatives in the near future including a command-line wizard, and a web-based input GUI.

Secondly, reproducibility is made easier through the use of 'profiles' that can define configuration parameters.  These profiles can be managed at different hierarchical levels. HPC-level profiles can specify parameters for the
computing environment (schedulers, cache locations, maximum resource etc.), which can be centrally managed to ensure all users of a group use the same settings. Pipeline-level profiles, specifying default parameters for
nf-core/eager itself, can be shared, thus making the pipeline portable between different groups through separation of the parameters from input data. Compared to the original EAGER that utilised per-FASTQ XML files with hardcoded filepaths, nf-core/eager allows researchers to publish the specific profile used in their runs alongside their publications, to ensure other groups can generate the same results. Usage of profiles also reduces mistakes caused by insufficient 'prose' based reporting of program settings, which may occur in papers written by researchers unfamiliar with informatics. The default nf-core/eager profile uses parameters evaluated in different aDNA specific contexts (e.g. @doi:10.1186/1471-2164-13-178), and will be updated in each new release as new studies are published.

nf-core/eager provides improved efficiency over the original EAGER pipeline by replacing the sample-by-sample sequential processing with Nextflow's asynchronous parallelisation. This, combined with pre-defined
per-process customisation of resource parameters, reduces unnecessary resource allocation that can occur with new users to each step of an NGS data processing pipeline. This is particularly pertinent given the increasing use of
centralised HPCs or cloud computing, which often use per-hour cost calculations.

## Latest aDNA practices

nf-core/eager follows a similar structural foundation with the original EAGER.
Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file.

This can be split into four main stages:

**ADD CITATIONS**

1. Preprocessing
   1. Sequencing quality control (FastQC @https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
   2. Sequencing artefact clean-up (AdapterRemoval2 (@doi:10.1186/s13104-016-1900-2))
   3. Preprocessing statistics generation
2. Mapping and post- processing
   1. Alignment against reference genome (BWA (@doi:10.1093/bioinformatics/btp324, @doi:10.1093/bioinformatics/btp698), CircularMapper(@doi:10.1186/s13059-016-0918-z))
   2. Mapping quality filtering (Samtools (@doi:10.1093/bioinformatics/btp352))
   3. PCR duplicate removal (DeDup(@doi:10.1186/s13059-016-0918-z), Picard MarkDuplicates(@doi:10.1101/gr.107524.110))
   4. Mapping statistics generation (PreSeq(@doi:10.1038/nmeth.2375), Qualimap2(@doi:10.1093/bioinformatics/btv566))
3. aDNA Evaluation and Modification
   1. Damage profiling (DamageProfiler(@doi:10.5281/zenodo.3557708))
   2. aDNA reads selection (PMDTools(@doi:10.1073/pnas.1318934111))
   3. Damage removal (Bamutils(@doi:10.1101/gr.176552.114))
   4. (Human) contamination estimation (ANGSD(@doi:10.1186/s12859-014-0356-4))
4. Genotyping and Consensus Sequencing (GATK(@doi:10.1101/gr.107524.110), VCF2Genome(@doi:10.1186/s13059-016-0918-z))

In nf-core/eager, all tools also originally used in EAGER have been updated
to latest compatible versions, as available on conda to ensure widespread
accessibility and stability. **Modules that have not been carried over are
Schmutzi and MapDamage because ALEX TO WRITE SOMETHING NICE HERE**. Support
for Bowtie2 will be added in the near future, after consultation with the
nf-core user community. New tools to the basic workflow include fastp
(@doi:10.1093/bioinformatics/bty560) for the removal of poly-G sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms). We have also included the
FreeBayes genotyper as an alternative to the human-focused GATK tools. We have
also maintained the possibility of using the now unsupported GATK 
UnifiedGenotyper, as the GATK HaplotypeCaller performs _de novo_ assembly
around possible variants, which may not be suitable for low-coverage aDNA data.  

We have further extended functionality of the pipeline, primarily focusing on
allowing for ancient metagenomic analysis to be run alongside genomic of the
reference genome. We have added the ability to screen all off-target reads
from mapping to the supplied reference FASTA with two metagenomic profilers:
MALT (@doi:10.1101/050559, @doi:10.1038/s41559-017-0446-6) and Kraken2
(@doi:10.1186/s13059-019-1891-0). Characterisation of properties of authentic
ancient DNA (@doi:10.1073/pnas.0704665104) from MALT alignments is carried out with the HOPS pipeline
(@doi:10.1186/s13059-019-1903-0). The pipeline also integrates a SNP alignment
generation tool, MultiVCFAnalyzer (doi:10.1038/nature13591), which allows
assessment of level of cross-mapping to a reference genome from different
related taxa - a common challenge in ancient bacteria genome reconstruction
(@doi:10.1146/annurev-genom-091416-035526). Simple coverage statistics of
particular annotations (e.g. genes) of an input reference  is offered by bedtools (@doi:10.1093/bioinformatics/btq033).

New tools supplied with the pipeline include a dedicated 'endogenous DNA'
calculator (endorS.py, by A.A.V), to give approximate level of
on-target DNA yield within a sample. Secondly, ancient metagenomic studies may
include comparative samples from living day individuals
(@doi:10.1186/s40168-019-0717-3). To support open data, whilst respecting data
privacy, nf-core/eager includes a 'strip_fastq' script (by M.B.) which creates
raw FASTQ files but with reference-genome mapped reads removed. This then
allows safe upload of sequencing data to public repositories with identifiable
data removed. When using a human reference genome, nf-core/eager also can give
estimates of the biological sex of a given individual with Sex.DetEERRmine
(@doi:10.1038/s41467-018-07483-5).

Complex sequencing strategies is also now accounted for with nf-core/eager.
Given the large amount of sequencing often required to yield sufficient genome
coverage from ancient DNA data, palaeogeneticists can sometimes use multiple
(differently treated) libraries or sequencing runs. As an alternative to direct
paths, the pipeline can also accept a TSV file which includes file paths and
additional metadata such as sample name, library name, sequencing  lane, colour
chemistry and UDG treatment. This allows simultaneous processing and appropriate
merging of heterogeneous data of multiple sequencing runs and/or libraries
types.

Finally, the original EAGER tabular report format has been replaced with MultiQC
(@doi:10.1093/bioinformatics/btw354). The original EAGER pipeline required users
to look through many independent output directories and files to make full
assessment of their sequencing data. Aggregation of all log files into a single
interactive report will assist users in making fuller assessment of their
sequencing and analysis runs. Most tools within nf-core/eager have a
corresponding MultiQC module to ensure as complete evaluation as possible.

## Accessibility

Alongside the portable new pipeline report, we have written extensive documentation
on all parts of running and interpreting the output of the pipeline. Given that
a large fraction of aDNA researchers come from fields such as social sciences
that have limited computational training (such as J.A.F.Y), we have written
documentation that also gives guidance on how to interpret each section of the
report, specifically in the context of NGS and aDNA. This includes schematic
images (by Z.F.) of best practices or expected output, published under CC-BY
licenses to allow for use in other training material. This open-access resource
will make the aDNA discipline more accessible to researchers new to the field,
by providing practical and 'applied' knowledge to how aDNA characteristics
translate to downstream analyses.

The development of nf-core/eager in nextflow and the nf-core initiative will
also improve open-source community contributions to the pipeline. While nextflow
is written primarily in the language groovy, the nextflow DSL simplifies a lot
of concepts to a level intermediate level that bioinformaticians without Java/Groovy
experience can easily access (regardless of own programming language experience).
Furthermore, Nextflow places ubiquitous command-line interfaces, such as bash,
in a prominent position within the code, rather than custom java code and
classes. This is particularly relevant for researchers ancient DNA, who often
come from social sciences and do not have backgrounds or training in
computational biology. We hope this will motivate further bug fixes and feature
contributions to keep the updated pipeline with standard practises during a
longer life-cycle of this version of the pipeline. This will also be supported
by the active and welcoming nf-core community who provide general guidance and
advice on developing nextflow and nf-core pipelines.
