## Methods

### Installation

nf-core/eager requires a version of Java, Nextflow and either a functional Conda
installation *or* Docker/Singularity container installation. A quick
installation guide to follow to get started can be found in the *Quickstart*
section of the nf-core/eager repository
[@url:https://github.com/nf-core/eager#quick-start].

### Running

After the installation, users can run the pipeline using standard test data by
utilizing some of the `test` profiles we provide (e.g. using Docker):

```bash
nextflow run nf-core/eager -r 2.1.0 -profile test,docker
```

This will download test data automatically, run the pipeline locally with all
software tools containerized in a Docker image and store the output of that run
in the `./results` folder of your current directory.

The default pipeline settings assumes paired end FASTQ data and will run:
FastQC; AdapterRemoval2 (merging and adapter clipping); post-clipping FastQC
(for AdapterRemoval2 performance evaluation); bwa mapping (with the 'aln'
algorithm); samtools flagstat (for mapping statistics); endorS.py (for
endogenous DNA calculation); DeDup (for PCR amplicon deduplication); PreSeq (for
library complexity evaluation); DamageProfiler and Qualimap2 (for genome
coverage statistics) and the MultiQC pipeline run report. If no additional FASTA
indices are given, these will also be generated.

The pipeline is highly configurable and most modules can be turned on and off at
the request of the user using different flags to allow high customisation to
each users needs. For example, to include metagenomic screening of off-target
reads and sex determination based on on-target mappings of pre-clipped
single-end data:

```bash
nextflow run nf-core/eager -r 2.1.0 -profile conda --input '/<path>/<to>/*/*R1*.fastq.gz' --single_end --fasta '/<path>/<to>/<reference>.fasta.gz' --skip_fastqc --skip_adapterremoval --run_bam_filtering --bam_discard_unmapped --bam_unmapped_type 'fastq' --run_metagenomic_screening --metagenomic_tool 'malt' --database '/<path>/<to>/<malt_database>' --run_sexdeterrmine
```

#### Profiles

We utilize a central configuration profile repository to enable users from
various institutions to use pipelines on their particular infrastructure more
easily [@url:https://github.com/nf-core/configs]. There are multiple resources
listed in this repository with information on how to add your own configuration
profile with help from the nf-core community.

Users can customize this infrastructure profile by themselves, with the nf-core
community, or with their local system administrator to make sure that the
pipeline runs successfully, and can then rely on the Nextflow and nf-core
framework to ensure compatibility upon further infrastructure changes. For
example, in order to run the nf-core/eager pipeline at the Max Planck Institute
for the Science of Human History (MPI-SHH) in Jena, users only have to run:

```bash
nextflow run nf-core/eager -r 2.1.0 -profile shh_cdag --input '/<path>/<to>/*/*{R1,R2}*.fastq.gz' --fasta '/<path>/<to>/<reference>.fasta.gz'
```

This runs the testing profile of the nf-core/eager pipeline with parameters
specifically adapted to the HPC system at the MPI-SHH. In some cases, similar
institutional configs for other institutions may already exist (originally
utilised for different nf-core pipelines), so users need not write their own.

#### Inputs

The pipeline can be started using (raw) FASTQ files from sequencing or
pre-mapped BAM files. Additionally, the pipeline requires a FASTA reference
genome.

If BAM input is provided, an optional conversion to FASTQ is offered, otherwise
BAM files processing will start from the post-mapping stage.

If users have complex set-ups, e.g. multiple sequencing lanes that require
merging of files for example, the pipeline can be supplied with a tabular
separated value (TSV) file to enable such complex data handling. Both FASTQs and
BAMs can be provided in this set up. FASTQs with the same library name and
sequencing chemistry but sequenced across multiple lanes will be concatenated
after AdapterRemoval and prior mapping. Libraries with the sample name and with
the same UDG treatment, will be merged after deduplication. If libraries with
the sample name have different UDG treatment, these will be merged after the
aDNA modification stage (i.e. BAM trimming or PMDtools, if turned on), prior
genotyping, as showin in Figure {@fig:merging-files}.

![Schematic of different processing and merging points based on the nature of
different libraries, as specified in metadata of a TSV file. Dashed boxes
represent optional library-specific
processes](images/Fig3_merging_files.png){#fig:merging-files width="70%"}

As Nextflow will automatically download files from URLs, profiles and/or TSV
files can include links to publicly available data (e.g. the ENA FTP server).
This assists in reproducibility as if profiles or TSV files are uploaded with a
publication, a researcher wishing to re-analyse the data in the same way can use
the exact settings and merging procedures in the original publication, without
having to reconstruct this from prose.

#### Monitoring

Users can either monitor their pipeline execution with the messages Nextflow
prints to the console while running, or utilize projects such as Nextflow Tower
[@url:https://tower.nf] to monitor their analysis pipeline during runtime.

#### Output

The pipeline produces several dozen output files in various file formats, with a
more detailed listing available in the user documentation. This includes
metrics, statistical analysis data and standardized output files (BAM, VCF) for
close inspection and further downstream analysis, as well as a MultiQC report.
If an emailing daemon is set up on the server, the latter can even be emailed to
users automatically, when starting the pipeline with a dedicated option (--email
you@yourdomain.org).

## Data and software availability

All code is available on github at
[https://github.com/nf-core/eager](https://github.com/nf-core/eager) and
archived with Zenodo under the DOI
[10.5281/zenodo.1465061](https://doi.org/10.5281/zenodo.1465061). All test data
is from the ENA public repository available under ENA IDs: **FIXME**

This paper was written with Manubot [@doi:10.1371/journal.pcbi.1007128].